### DISCUSSION TITLE
Eliminating need for resource directory/builtin headers

### DISCUSSION THREAD

**Discussion by sam-mccall** [clangd/clangd#970] [Ideas]
(I've investigated this and I'm not sure it's worth the build times, maybe others have ideas of how to improve or opinions).

At the moment clangd (and other tools) require the built-in headers (`stddef.h` and friends). clangd searches for them relative to the binary in `../lib/clang/<version>/include`, so we include them in the zip bundle with the right structure.

This adds a failure mode - people are sometimes surprised when they move the clangd binary and it stops working. It's also just ugly and another weird thing to remember. I'd love to get rid of it.

The most obvious option is to embed the file contents in the binary and inject them into the VFS used to compile. You can't navigate to the file in your editor, but I think we could live with that. The bigger problem is the build system.
 - the fastest way to embed the files is for something to turn them into object files directly, like `objcopy -I binary`, but this isn't portable. I guess we'd end up maintaining a mess of platform-specific cmake nonsense.
 - the portable way is to generate C++ source code containing arrays of bytes, and then build those. But this has scaling problems when there's a lot of data.

I thought of these as "a few small headers" but there are also processor-specific "intrinsics" headers like `avx2intrin.h`, `arm_neon.h` etc. These are often table-generated, describing hundreds of instruction/data type variants, and are the bulk of the total size. Of 20MB total headers, `riscv_vector.h` is 8MB(!) and `arm_*.h` are 5.5MB, and there's PPC and various x86 things too.
So we could expect binary size to increase by 20MB, though we're bundling that data anyway.

My first attempt was to embed the contents in C++ rawstrings, which would at least be fast to parse. Unfortunately compilers enforce a 64k limit on string literals, for GCC/clang this is a warning but in MSVC it's a hard error.

Next attempt was `const unsigned char data[] = {64,65,66,...};`
At these data sizes it's slow to parse: 27seconds of CPU and 2GB of ram on a powerful machine for everything. (We can split up each file into a separate TUs to parallelize, but remember the largest single file is 40% of the total).
This was with clang-11 and I'm concerned some supported compilers might be significantly slower.

Anyone have a better technique? Thoughts on whether slowing down the build for this is worth it?

**Comment by HighCommander4** [comment]
Anecdotally, based on how often I've seen users run into this and ask questions (on Discord/Discourse/Github), this is dwarfed by other configuration issues. I'd put #539 much higher on the list than this.
